kas/anchor.go

```diff
@@ -0,0 +1,126 @@
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package kas
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/common"
+	"net/http"
+)
+
+var (
+	codeOK                = 0
+	errNotFoundBlock      = errors.New("not found block")
+	errInvalidBlockNumber = errors.New("invalid block number")
+)
+
+//go:generate mockgen -destination=./mocks/anchordb_mock.go -package=mocks github.com/klaytn/klaytn/kas AnchorDB
+type AnchorDB interface {
+	WriteAnchoredBlockNumber(blockNum uint64)
+	ReadAnchoredBlockNumber() uint64
+}
+
+//go:generate mockgen -destination=./mocks/blockchain_mock.go -package=mocks github.com/klaytn/klaytn/kas BlockChain
+type BlockChain interface {
+	GetBlockByNumber(number uint64) *types.Block
+}
+
+//go:generate mockgen -destination=./mocks/client_mock.go -package=mocks github.com/klaytn/klaytn/kas HTTPClient
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type Anchor struct {
+	kasConfig *KASConfig
+	db        AnchorDB
+	bc        BlockChain
+	client    HTTPClient
+}
+
+func NewKASAnchor(kasConfig *KASConfig, db AnchorDB, bc BlockChain) *Anchor {
+	return &Anchor{
+		kasConfig: kasConfig,
+		db:        db,
+		bc:        bc,
+		client:    &http.Client{},
+	}
+}
+
+type respBody struct {
+	Code   int         `json:"code"`
+	Result interface{} `json:"result"`
+}
+
+type reqBody struct {
+	Operator common.Address `json:"operator"`
+	Payload  interface{}    `json:"Payload"`
+}
+
+type Payload struct {
+	Id string `json:"id"`
+	types.AnchoringDataInternalType0
+}
+
+func dataToPayload(anchorData *types.AnchoringDataInternalType0) *Payload {
+	payload := &Payload{
+		Id:                         anchorData.BlockNumber.String(),
+		AnchoringDataInternalType0: *anchorData,
+	}
+
+	return payload
+}
+func (anchor *Anchor) sendRequest(payload interface{}) (*respBody, error) {
+	header := map[string]string{
+		"Content-Type": "application/json",
+		"X-Krn":        anchor.kasConfig.Xkrn,
+	}
+
+	bodyData := reqBody{
+		Operator: anchor.kasConfig.Operator,
+		Payload:  payload,
+	}
+
+	bodyDataBytes, err := json.Marshal(bodyData)
+	if err != nil {
+		return nil, err
+	}
+
+	body := bytes.NewReader(bodyDataBytes)
+
+	req, err := http.NewRequest("POST", anchor.kasConfig.Url, body)
+	if err != nil {
+		return nil, err
+	}
+	req.SetBasicAuth(anchor.kasConfig.User, anchor.kasConfig.Pwd)
+	for k, v := range header {
+		req.Header.Set(k, v)
+	}
+
+	resp, err := anchor.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```

Comment:
no need to check `resp.Body != nil`?

Reply:
It is not necessary.
See here.

https://golang.org/src/net/http/response.go
```
	// The http Client and Transport guarantee that Body is always
	// non-nil, even on responses without a body or responses with
	// a zero-length body. It is the caller's responsibility to
	// close Body. The default HTTP client's Transport may not
	// reuse HTTP/1.x "keep-alive" TCP connections if the Body is
	// not read to completion and closed.
```